### PR TITLE
Integration methods and classes for deeptrack

### PR DESCRIPTION
**Pull request by BenjaminMidtvedt**
Introduces many convenience integration methods to the `Application` class (and its subclasses) for integrating with deeptrack pipelines, but also just general convenience. There are some pretty major design decisions here, so we should all be on the same page.

# Major decision, automatic conversion to channel first

From now, it will be possible to directly pass `np.ndarrays` and `deeptrack.Feature`s to some deeplay methods. 

1. `deeplay` will assume that `numpy.ndarrays` either given directly or produced by deeptrack features are _channel last_, and automatically convert them to channel first. We can, if this proves troublesome in the future, give to option to disable this behavior.
2. Rule 1. does not apply if the data type is `integer`. Why? Because torch decided that CrossEntropyLoss targets should have no channels. To my knowledge, there is no `float` loss that does not match the shape of `y_hat`. 
3. `torch.Tensor`s will not be permuted. So, if one does the work themselves to convert to ttorch tensors, you will not run into any unexpected conversions.

Why do we need to do this? We don't, really. But it i highly unintuitive that to packages that are designed to work together requires the user to do manual permuting of channels back and forth to work. 

Moreover, we attempt to do automatic dtype conversions if possible. In general, any float will be converted to the default torch float, and any int will be converted to `long`. Finally, because `torch` decided to make things hard, for `BCEloss` and `BCELossWithLogits` we automatically convert the targets to `float`. 

The idea is not to always be correct, but to, in the cases we know what the correct dtype is, we perform the conversion. If we are not sure, we leave it to the user.

Since all of this only applies to non-tensors, no existing code should be affected.

# New shiny things!

With that out of the way, let's go through the new methods.

## `Application.fit`

This is a keras-like fit method for maximum convenience. Here's an example:

```py
model = dl.CategoricalClassifier(net)
model.fit(
    train_data,
    max_epochs=100,
    batch_size=32,
    val_data=val_data,
    val_batch_size=256
)
```

Here, train_data and val_data can be any of:
- DataLoader
- torch Dataset
- deeptrack Dataset
- deeptrack Feature (pipeline)
- tuple of ndarrays or tensors

For Feature's, specifically, you can provide the additional kwargs
- `steps_per_epoch`
- `val_steps_per_epoch`
since features are unsized.

All additional kwargs will be passed to the trainer, so you can still set conventional trainer kwargs like `callbacks`, `accelerator` etc. Moreover, it uses the rich progressbar, which looks good on all platforms, unlike the nbwidget one which is awful in vscode.

## `LogHistory`

Missed the history object from keras? It's back! But better, kinda. It will automatically be returned by `Application.fit`, but you can also just provide it as a `callback` to the trainer. It has two properties and a plot method. For most cases, it should be enough to access the plot method:

image

A nice little plot made with the LogHistory object.

You can also access `h.history` and `h.step_history`. They contain the `on_epoch` metrics and the `on_step` metrics respectively. They are given in the format 
```py
{
   metric_name: {
       "value": [0.6, 0.5, 0.4, ...],
       "step": [0, 100, 200, ...],
       "epoch": [0, 1, 2, ...]
   }
}
```

## `Application.test`

A frustration for me was that you could only do `trainer.test` for metrics that you defined upon creating the model with `metrics=[...]`. With this, you can test any metrics:

```py
f1 = tm.F1Score(task="binary")
confusion_matrix = tm.ConfusionMatrix(task="binary")
ROC = tm.ROC(task="binary")

results = model.test(test, metrics={"f1": f1, "confusion_matrix": confusion_matrix, "roc": ROC})
```

Here, metrics can be provided in any of the following formats, but will always be returned as a dict:
- dict
- tm.Metric (output key is .__name__)
- tuple (key, metric)
- a list of the previous two

### CODE REVIEW

**Comment by JesusPinedaC** [comment]
This sounds very nice to me!

**Comment by JesusPinedaC** [comment]
Loading the net's weights from a checkpoint would still works in the standard way, right? e.g., `model.load_from_checkpoint(path, model=net)`

**Comment by BenjaminMidtvedt** [comment]
@JesusPinedaC That's the idea yes. I've verified that loading from a state dict works. My hope is that I can implement it such that the config object gets saved with the checkpoint. That way you can load the entire structure of the network directly, but this is not yet implemented.

**Comment by giovannivolpe** [comment]
It seems good also to me
